packages/arb-ts/src/lib/utils/arbProvider.ts

```diff
@@ -0,0 +1,246 @@
+    const hash = this.hash.bind(this)
+    const number = this.number.bind(this)
+    const feeStats = this.feeStats.bind(this)
+    const feeStat = this.feeStat.bind(this)
+    const batchInfo = this.batchInfo.bind(this)
+    const returnCode = this.returnCode.bind(this)
+
+    const arbTransactionFormat = {
+      ...superFormats.transaction,
+
+      l1SequenceNumber: bigNumber,
+      parentRequestId: hash,
+      indexInParent: number,
+      arbType: number,
+      arbSubType: number,
+      l1BlockNumber: number,
+    }
+
+    const arbReceiptFormat = {
+      ...superFormats.receipt,
+      returnData: data,
+      returnCode: returnCode,
+      feeStats: feeStats,
+      batchInfo: Formatter.allowNull(batchInfo, null),
+      l1BlockNumber: number,
+    }
+
+    const feeStatsFormat = {
+      prices: feeStat,
+      unitsUsed: feeStat,
+      paid: feeStat,
+    }
+
+    const feeStatFormat = {
+      l1Transaction: bigNumber,
+      l1Calldata: bigNumber,
+      l2Storage: bigNumber,
+      l2Computation: bigNumber,
+    }
+
+    const batchInfoFormat = {
+      confirmations: number,
+      blockNumber: number,
+      logAddress: address,
+      logTopics: Formatter.arrayOf(hash),
+      logData: data,
+    }
+
+    return {
+      ...superFormats,
+      transaction: arbTransactionFormat,
+      receipt: arbReceiptFormat,
+      feeStats: feeStatsFormat,
+      feeStat: feeStatFormat,
+      batchInfo: batchInfoFormat,
+    }
+  }
+
+  public returnCode(value: any): ReturnCode {
+    const bn = BigNumber.from(value)
+    const returnNum = bn.toNumber()
+    if (!Object.values(ReturnCode).includes(returnNum)) {
+      return logger.throwArgumentError('invalid return code', 'value', value)
+    }
+    return returnNum
+  }
+
+  public feeStat(feeStat: any): FeeStat {
+    return Formatter.check(this.formats.feeStat, feeStat)
+  }
+
+  public feeStats(feeStats: any): FeeStats {
+    return Formatter.check(this.formats.feeStats, feeStats)
+  }
+
+  public batchInfo(batchInfo: any): BatchInfo {
+    return Formatter.check(this.formats.batchInfo, batchInfo)
+  }
+
+  public transactionResponse(transaction: any): ArbTransactionResponse {
+    return super.transactionResponse(transaction) as ArbTransactionResponse
+  }
+
+  public receipt(value: any): ArbTransactionReceipt {
+    return super.receipt(value) as ArbTransactionReceipt
+  }
+}
+
+const getBatch = async (
+  seqNum: BigNumber,
+  l1BlockNumber: number,
+  l1Provider: providers.Provider
+): Promise<Omit<BatchInfo, 'confirmations'> | null> => {
+  const ethNetwork = await getL1Network(l1Provider)
+  const arbNetwork = await getL2Network(ethNetwork.partnerChainIDs[0])
+  const inbox = SequencerInbox__factory.connect(
+    arbNetwork.ethBridge.sequencerInbox,
+    l1Provider
+  )
+
+  const delta = 1000
+
+  const delayBlocks = (await inbox.maxDelayBlocks()).toNumber()
+
+  const startBlock = l1BlockNumber - delta
+  const delayedBlockMax = l1BlockNumber + delayBlocks
```

Comment:
This upper bound only applies to transactions submitted in batches by the sequencer ( they indeed can't be included after `l1BlockNumber + delayBlocks`). Transactions that are force-included, however, can be force-included after any arbitrary amount of time. So since this `getBatch`method wants to cover `DelayedInboxForcedEvent`, as is, it would have to search past `l1BlockNumber + delayBlocks`.

Reply:
There's a few options here: one is to do the initial event log query only on `SequencerBatchDelivered` and `SequencerBatchDelivered` (using the `delayedBlockMax ` var as is), and then iff those return nil, _and_ the current L1 block is past the point of l1BlockNumber + delayBlocks, we do a second query for `DelayedInboxForced`. Unlovely to have to do 2 event queries, but  the 2nd query actually would only happen for a very unusual case.